packages/token/src/tests/erc721/test_erc721_enumerable.cairo

```diff
@@ -0,0 +1,581 @@
+
+    // Note that the original last indexed token id is now the first because of the
+    // swap-and-pop operation.
+    assert_owner_tokens_index_to_id(OWNER(), first_token_index, last_token_id);
+    assert_owner_tokens_id_to_index(first_token_id, 0);
+}
+
+//
+// _remove_token_from_all_tokens_enumeration
+//
+
+#[test]
+fn test__remove_token_from_all_tokens_enumeration_with_last_token() {
+    let (mut state, tokens_list) = setup();
+    let initial_supply = state.total_supply();
+    let last_token_index = state.total_supply() - 1;
+    let last_token_id = *tokens_list.at(last_token_index.try_into().unwrap());
+
+    assert_all_tokens_index_to_id(last_token_index, last_token_id);
+    assert_all_tokens_id_to_index(last_token_id, last_token_index);
+
+    state._remove_token_from_all_tokens_enumeration(last_token_id);
+
+    assert_all_tokens_index_to_id(last_token_index, last_token_id);
+    assert_all_tokens_id_to_index(last_token_id, last_token_index);
+
+    // Check supply
+    let new_supply = state.total_supply();
+    assert_eq!(initial_supply - 1, new_supply);
+}
+
+#[test]
+fn test__remove_token_from_all_tokens_enumeration_with_first_token() {
+    let (mut state, tokens_list) = setup();
+    let initial_supply = state.total_supply();
+    let first_token_index = 0;
+    let first_token_id = *tokens_list.at(0);
+    let last_token_index = tokens_list.len() - 1;
+    let last_token_id = *tokens_list.at(last_token_index);
+
+    assert_all_tokens_index_to_id(first_token_index, first_token_id);
+    assert_all_tokens_id_to_index(first_token_id, first_token_index);
+
+    state._remove_token_from_all_tokens_enumeration(first_token_id);
+
+    assert_all_tokens_index_to_id(first_token_index, last_token_id);
+    assert_all_tokens_id_to_index(first_token_id, 0);
+
+    // Check supply
+    let new_supply = state.total_supply();
+    assert_eq!(initial_supply - 1, new_supply);
+}
+
+//
+// Helpers
+//
+
+fn assert_dual_token_of_owner_by_index(owner: ContractAddress, expected_token_list: Span<u256>) {
+    let mut state = COMPONENT_STATE();
+
+    let mut i = 0;
+    while i < expected_token_list.len() {
+        let token = state.token_of_owner_by_index(owner, i.into());
+        assert_eq!(token, *expected_token_list.at(i));
+
+        i += 1;
+    }
+}
+
+fn assert_dual_token_by_index(expected_token_list: Span<u256>) {
+    let mut state = COMPONENT_STATE();
+
+    let mut i = 0;
+    while i < expected_token_list.len() {
+        let token = state.token_by_index(i.into());
+        assert_eq!(token, *expected_token_list.at(i));
+
+        i += 1;
+    }
+}
+
+fn assert_after_update_all_tokens_list(expected_list: Span<u256>) {
+    let state = @COMPONENT_STATE();
+
+    let mut i = 0;
+    while i < expected_list.len() {
+        // Check total tokens list
+        let token = state.ERC721Enumerable_all_tokens.read(i.into());
+        assert_eq!(token, *expected_list.at(i));
+
+        i += 1;
+    }
+}
```

Comment:
It would be helpful to implement helper functions to fetch tokens using ERC721Enumerable functionality and then to assert that the result list of tokens is equal to the expected one

```
fn fetch_tokens_by_owner(owner: ContractAddress) -> Span<u256> {
    let contract = CONTRACT_STATE();
    let tokens_len = contract.balance_of(owner);
    let mut i = 0;
    let mut result = array![];
    while i < tokens_len {
        let token_id = contract.token_of_owner_by_index(owner, i);
        result.append(token_id);
        i += 1;
    };
    result.span()
}

fn fetch_all_tokens() -> Span<u256> {
    let contract = CONTRACT_STATE();
    let tokens_len = contract.total_supply();
    let mut i = 0;
    let mut result = array![];
    while i < tokens_len {
        let token_id = contract.token_by_index(i);
        result.append(token_id);
        i += 1;
    };
    result.span()
}
```

Reply:
This would make the assertions a lot cleaner by abstracting out the loops! What I don't like about it though is that it's abstracting out the function call that's being asserted (`token_by_index` and `token_of_owner_by_index`).

One of the loops also needs to read from storage directly for `before_update` tests in order to bypass the out of bounds check. I added comments for clarity. That said, I don't think it's really justified to abstract the assertions further. There's only three loops in the helpers. Let me know what you think and if you disagree

Reply:
I checked the new assertion helpers and completely happy with the implementation 👍